services/verification-data.js

```diff
@@ -0,0 +1,101 @@
+import { EmailVerificationModel } from "../schemas/email-verification.js";
+import dayjs from "dayjs";
+
+/**
+ * This method handles the logic for sending the verification code e-mail
+ * When to send it, generating codes, marking if the verification was complete
+ * @param {string} email
+ * @returns {Promise<{ _id: any, email: string, code: string, verified: boolean, created: Date, expires: Date, createdAt: Date, updatedAt: Date}> | Promise<null>}
+ */
+export async function createVerificationData (email) {
+	const verificationDocument = await EmailVerificationModel.findOne({ email: email }).exec();
+	if (!verificationDocument) {
+		// Create an entry
+		return createNewVerificationEntry(email)
+	}
+
+	// Verified e-mails don't need to be verified again
+	if (verificationDocument.verified === true) return null;
+
+	// Let's give this a 20 minute cooling off period so we don't send too many requests just in case	
+	if (dayjs().isBefore(dayjs(verificationDocument.created).add(20, 'minute'))) {
+		console.log(new Date().toISOString())
+		throw new Error("Verification code cool off period");
+	}
+
+	// If one already exists and is not expired, just return it
+	if (!isExpired(verificationDocument.expires)) {
+		return verificationDocument
+	} else {
+		// If it's expired, update
+		return refreshRequest(verificationDocument);
+	}
+}
+
+/**
+ * 
+ * @param {string} email 
+ * @returns {Promise<boolean>}
+ */
+export async function isEmailVerified(email) {
+	const verificationDocument = await EmailVerificationModel.findOne({ email: email }).exec();
+	return verificationDocument?.verified === true;
+}
+
+async function createNewVerificationEntry(email) {
+	return EmailVerificationModel.create({
+		email: email,
+		code: generateVerificationCode(),
+		created: new Date(),
+		expires: dayjs().add(20, 'minute').toDate()
+	})
+}
+
+/**
+ * 
+ */
+function generateVerificationCode() {
+	return Math.floor(100000 + Math.random() * 900000).toString();
```

Comment:
I like this. One issue, though: someone may try to try several random codes in quick sucession , as if to brute-force something. This is good for practical and demonstration purposes but there should be a rule, like: first 8 characters are random, the next 8 characters are a calculated hash based on the first 8. This is to prevent someone from depleting DB read quota.